waiter/src/waiter/auth/saml.clj

```diff
@@ -0,0 +1,379 @@
+            [ring.middleware.params :as ring-params]
+            [ring.util.codec :refer [form-encode url-encode base64-encode]]
+            [ring.util.response :refer [redirect]]
+            [waiter.auth.authentication :as auth]
+            [waiter.middleware :as middleware]
+            [waiter.util.utils :as utils])
+  (:import [javax.xml.parsers DocumentBuilderFactory]
+           [org.opensaml.xml.validation ValidationException]))
+
+(def charset-format (java.nio.charset.Charset/forName "UTF-8"))
+
+(defn read-to-end
+  [stream]
+  (let [sb (StringBuilder.)]
+    (with-open [reader (-> stream
+                         java.io.InputStreamReader.
+                         java.io.BufferedReader.)]
+      (loop [c (.read reader)]
+        (if (neg? c)
+          (str sb)
+          (do
+            (.append sb (char c))
+            (recur (.read reader))))))))
+
+(defn str->bytes
+  [some-string]
+  (.getBytes some-string charset-format))
+
+(defn bytes->str
+  [some-bytes]
+  (String. some-bytes charset-format))
+
+(defn byte-deflate
+  [str-bytes]
+  (let [out (java.io.ByteArrayOutputStream.)
+        deflater (java.util.zip.DeflaterOutputStream.
+                   out
+                   (java.util.zip.Deflater. -1 true) 1024)]
+    (.write deflater str-bytes)
+    (.close deflater)
+    (.toByteArray out)))
+
+(defn str->deflate->base64
+  [deflatable-str]
+  (let [byte-str (str->bytes deflatable-str)]
+    (bytes->str (b64/encode (byte-deflate byte-str)))))
+
+(defn base64->str
+  [string]
+  (let [byte-str (str->bytes string)]
+    (bytes->str (b64/decode byte-str))))
+
+(defn uri-query-str
+  [clean-hash]
+  (form-encode clean-hash))
+
+(defn get-idp-redirect
+  "Return Ring response for HTTP 302 redirect."
+  [idp-url saml-request relay-state]
+  (redirect
+    (str idp-url
+         "?"
+         (let [saml-request (str->deflate->base64 saml-request)]
+           (uri-query-str
+             {:SAMLRequest saml-request :RelayState relay-state})))))
+
+(defn certificate-x509
+  "Takes in a raw X.509 certificate string, parses it, and creates a Java certificate."
+  [x509-string]
+  (let [certificate-factory (java.security.cert.CertificateFactory/getInstance "X.509")
+        certificate-input-stream (io/input-stream (.getBytes x509-string))]
+    (.generateCertificate certificate-factory certificate-input-stream)))
+
+(defn jcert->public-key
+  "Extracts a public key object from a java cert object."
+  [java-cert-obj]
+  (.getPublicKey java-cert-obj))
+
+(defn saml-assertion-signature-valid?
+  "Checks that the SAML assertion has a valid signature."
+  [assertion idp-cert]
+  (when-let [signature (.getSignature assertion)]
+    (let [idp-pubkey (-> idp-cert certificate-x509 jcert->public-key)
+          public-creds (doto (new org.opensaml.xml.security.x509.BasicX509Credential)
+                         (.setPublicKey idp-pubkey))
+          validator (new org.opensaml.xml.signature.SignatureValidator public-creds)]
```

Comment:
Please use the simple class name in the body and move the fully qualified name to the import.

Reply:
done